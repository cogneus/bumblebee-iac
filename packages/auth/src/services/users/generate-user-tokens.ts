import { SecretsClient, ServiceContext, Token, ensureClient } from 'bumblebee-common'
import { getSigningKey } from '../signing'
import { getTokenServiceAttrs } from '../../helpers'
import { generateToken } from '../token'

export const generateUserTokens = async (serviceContext: ServiceContext) => {
  const secrets = ensureClient<SecretsClient>(serviceContext, 'secrets')
  const usersConfig = await secrets.getValue(secrets.ids.users.name)
  const key = await getSigningKey(serviceContext)
  const users: Token[] = usersConfig ? JSON.parse(usersConfig) : []
  const serviceAttrs = getTokenServiceAttrs()
  const promises = users.map((user) => generateToken(key, user, serviceAttrs))
  const tokenResponses = await Promise.all(promises)
  const tokens = tokenResponses.reduce((collection, { token }, index) => {
    const { subject } = users[index]
    return {
      ...collection,
      [subject]: token,
    }
  }, {})
  await secrets.setValue(`${secrets.names.tokens.name}`, tokens)
}
