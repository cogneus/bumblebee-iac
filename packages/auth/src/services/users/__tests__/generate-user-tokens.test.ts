import { ISecretsClient, SecretsClientMock, ServiceContext, Token } from 'bumblebee-common'
import fs from 'fs'
import path from 'path'
import { generateUserTokens } from '../generate-user-tokens'

describe('Generate user token tests', () => {
  beforeEach(() => {
    jest.clearAllMocks()
    jest.useFakeTimers()
    jest.setSystemTime(new Date('2023-01-01T00:00:00Z'))
    process.env = {
      AWS_REGION: 'eu-west-1',
      LOG_LEVEL: 'debug',
      NODE_ENV: 'dev',
      RESOURCE_NAME_PREFIX: 'test-prefix',
      SECRET_SIGNING_KEY_NAME: 'authSigningKey',
      TOKEN_AUDIENCE: 'audience',
      TOKEN_EXPIRY: '12h',
      TOKEN_ISSUER: 'issuer',
      SECRET_TOKENS_NAME: 'tokens',
      SSM_NAME_PREFIX: 'ssm-prefix',
      SSM_USERS_NAME: 'users',
    } as any
  })

  const file = path.join(__dirname, './', 'test-key.key')
  const key = fs.readFileSync(file, 'utf8')

  const secretsClient: ISecretsClient = new SecretsClientMock({
    prefix: 'test-prefix',
    idPrefix: 'test-prefix',
    names: {
      authSigningKey: { name: 'authSigningKey' },
      tokens: { name: 'tokens' },
      users: { name: 'users' },
    },
    ids: {
      authSigningKey: { name: 'authSigningKey' },
      tokens: { name: 'tokens' },
      users: { name: 'users' },
    },
    kmsId: 'test',
  })

  const users: Token[] = [
    {
      email: 'read-user@test.email',
      givenName: 'read',
      legalName: 'user',
      rights: ['read'],
      subject: 'read-user',
      uid: 'read-user',
    },
    {
      email: 'write-user@test.email',
      givenName: 'write',
      legalName: 'user',
      rights: ['write'],
      subject: 'write-user',
      uid: 'write-user',
    },
  ]
  test('should create user tokens', async () => {
    const results: any[] = []
    ;(secretsClient.getValue as jest.Mock).mockImplementation((name) =>
      name === secretsClient.names.authSigningKey.name
        ? key
        : name === secretsClient.names.users.name
        ? JSON.stringify(users)
        : null
    )
    ;(secretsClient.setValue as jest.Mock).mockImplementation((name, value) =>
      results.push({ name, value })
    )
    await generateUserTokens({
      clients: { secrets: secretsClient },
    } as ServiceContext)
    expect(results).toMatchSnapshot()
  })
})
