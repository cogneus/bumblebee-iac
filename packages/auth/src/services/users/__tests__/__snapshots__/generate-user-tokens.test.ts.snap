// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate user token tests should create user tokens 1`] = `
[
  {
    "name": "tokens",
    "value": {
      "read-user": "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6InJlYWQtdXNlckB0ZXN0LmVtYWlsIiwiZ2l2ZW5OYW1lIjoicmVhZCIsImxlZ2FsTmFtZSI6InVzZXIiLCJyaWdodHMiOlsicmVhZCJdLCJzdWJqZWN0IjoicmVhZC11c2VyIiwidWlkIjoicmVhZC11c2VyIiwiaWF0IjoxNjcyNTMxMjAwLCJleHAiOjE2NzI1NzQ0MDAsImF1ZCI6ImF1ZGllbmNlIiwiaXNzIjoiaXNzdWVyIn0.aLfaZo1ca-moXsqLliW-4dL-LY41vGGyF3MWnZdhK5CpumhYOihXi59Cx1Fi3vclShZqq61biuKicBRlYzOZOvrCjpoHSY_efpqqmggPTyBo6yzLxauuwcTrny5kUDtx6bTsixkAzfTi0M1tjL1lT0cvMwr5V3NaZcIGtQUSQ7ynZBksFbtAdIYTNdBL31jsmgFRhiwm0o01sDUzfLKX6is71_fhQLaGBj6FCD1Rx1HhqfYQAttXBOBWOzX0Gv6Co28ef3NUBiWCU-n23lKk_MuZvmxr2UkPdCeLJn6Rr7yDGUWujaTkTji0UIJKFeKIC4HN8cUejFMrWwQU8pExGA",
      "write-user": "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJlbWFpbCI6IndyaXRlLXVzZXJAdGVzdC5lbWFpbCIsImdpdmVuTmFtZSI6IndyaXRlIiwibGVnYWxOYW1lIjoidXNlciIsInJpZ2h0cyI6WyJ3cml0ZSJdLCJzdWJqZWN0Ijoid3JpdGUtdXNlciIsInVpZCI6IndyaXRlLXVzZXIiLCJpYXQiOjE2NzI1MzEyMDAsImV4cCI6MTY3MjU3NDQwMCwiYXVkIjoiYXVkaWVuY2UiLCJpc3MiOiJpc3N1ZXIifQ.lwKB0ykljqxCuo373xNf_h67GWwCB9XZ2H0qKHRXdbLRNjIjmzNJYxBwISRtqt1Igrp22BKoLKHy52lVbam5UNUTZFKCCZR3aueNc2u7RqA1H7iP4whY4hxRgGpaIjL6A7WYWHTQ5J7RzXyiuirLw1k2hdRdpApiLcs8LZMo5oG1LEovfBuUldIR9DA6_ZNJcrPvsZgh7AL1LYpBDx_h8ZOfgNW5NFPphcXMEO-c9nX1ICt6H_c_0WMs__QwDnOYp4IaMbrLfQpRH2f8bIJ5KZGkPPrTPD9TLle3e50hBMm3lTo13ZNTAd7gzYYRebF8AjVxZqUK16FEM2xosM78gg",
    },
  },
]
`;
